src/ol_infrastructure/components/services/vault.py

```diff
@@ -1,5 +1,104 @@
-import pulumi
-from pulumi_vault import database
+"""
+This module controls mounting and configuration of database secret backends in Vault.
 
+This includes:
+- Mount a database backend at the configured mount point
+- Set the configuration according to the requirements of the backend type
+- Control the lease TTL settings
+- Define the base set of roles according to our established best practices
+"""
+from enum import Enum
+from typing import Text
 
-class VaultDatabaseRole(pulumi.ComponentResource):
+from pulumi import ComponentResource, ResourceOptions
+from pulumi_vault import Mount, database
+from pydantic import BaseModel, SecretStr
+
+from ol_infrastructure.lib.vault import (
+    mysql_sql_statements,
+    postgres_sql_statements
+)
+
+DEFAULT_PORT_POSTGRES = 5432
+DEFAULT_PORT_MYSQL = 3307
+DEFAULT_POT_MONGODB = 27017
+SIX_MONTHS = 60 * 60 * 24 * 30 * 6
+
+class DBEngines(str, Enum):
+    """Constraints for valid engine types that are supported by this component."""
+    postgres = 'postgresql'
+    mariadb = 'mysql'
+    mysql = 'mysql'
+    mysql_rds = 'mysql_rds'
+    mongodb = 'mongodb'
+
+class OLVaultDatabaseConfig(BaseModel):
+    db_name: Text
+    mount_point: Text
+    db_admin_username: Text
+    db_admin_password: SecretStr
+    verify_connection: bool = True
+    db_host: Text
+    max_ttl: int = SIX_MONTHS
+    default_ttl: int = SIX_MONTHS
+
+
+class OLVaultPostgresDatabaseConfig(OLVaultDatabaseConfig):
+    db_port: Int = DEFAULT_PORT_POSTGRES
+    db_connection: Text = 'postgresql://{{db_user}}:{{db_pass}}@{{db_host}}:{{db_port}}/{{db_name}}'
+    db_type: Text = DBEngines.postgres.value
+    role_statements: Dict[Text, Text] = postgresql_sql_statements
+
+
+class OLVaultMysqlDatabaseConfig(OLVaultDatabaseConfig):
+    db_port: Int = DEFAULT_PORT_MYSQL
+    db_connection: Text = '{{db_user}}:{{db_pass}}@tcp({{db_host}}:{{db_port}})/'
+    db_type: Text = DBEngines.mysql_rds.value
+    role_statements: Dict[Text, Text] = postgresql_sql_statements
+
+
+class OLVaultDatabaseBackend(ComponentResource):
+
+    def __init__(self, db_config: OLVaultDatabaseConfig, opts: ResourceOptions = None):
+        super().__init__('ol:services:VaultDatabaseBackend', db_config.db_name, None, opts)
+
+        resource_opts = ResourceOptions(parent=self).merge(opts)
+
+        self.db_mount = Mount(
+            f'{db_config.db_name}-mount-point',
```

Comment:
Looks like in the class definition the [name](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_infrastructure/components/aws/database.py#L57) is optional. Do we want to change that in order to avoid having a db_mount that doesn't have a db_name?

Reply:
This is a different db_config object. The `OLVaultDatabaseConfig` has a `db_name` attribute that's required. For the RDS component, the db_name is for the schema to be added after creation of the instance.

Reply:
So we're setting up a separate vault mountpoint for each db as opposed to how we currently have some of the mount points defined where we use the same mountpoint but use different paths to split envs. Why did we choose to go that route and do we foresee any issues with this approach?

Reply:
We've always had different mount points for the DBs. The generic secret mounts are where we have different paths per env.